modules/optflow/src/pcaflow.cpp

```diff
@@ -0,0 +1,508 @@
+ // any express or implied warranties, including, but not limited to, the implied
+ // warranties of merchantability and fitness for a particular purpose are disclaimed.
+ // In no event shall the Intel Corporation or contributors be liable for any direct,
+ // indirect, incidental, special, exemplary, or consequential damages
+ // (including, but not limited to, procurement of substitute goods or services;
+ // loss of use, data, or profits; or business interruption) however caused
+ // and on any theory of liability, whether in contract, strict liability,
+ // or tort (including negligence or otherwise) arising in any way out of
+ // the use of this software, even if advised of the possibility of such damage.
+ //
+ //M*/
+
+#include "opencv2/ximgproc/edge_filter.hpp"
+#include "precomp.hpp"
+
+namespace cv
+{
+namespace optflow
+{
+
+OpticalFlowPCAFlow::OpticalFlowPCAFlow( Ptr<const PCAPrior> _prior, const Size _basisSize, float _sparseRate,
+                                        float _retainedCornersFraction, float _occlusionsThreshold,
+                                        float _dampingFactor )
+    : prior( _prior ), basisSize( _basisSize ), sparseRate( _sparseRate ),
+      retainedCornersFraction( _retainedCornersFraction ), occlusionsThreshold( _occlusionsThreshold ),
+      dampingFactor( _dampingFactor )
+{
+  CV_Assert( sparseRate > 0 && sparseRate <= 0.1 );
+  CV_Assert( retainedCornersFraction >= 0 && retainedCornersFraction <= 1.0 );
+  CV_Assert( occlusionsThreshold > 0 );
+}
+
+template <typename T> static inline int mathSign( T val ) { return ( T( 0 ) < val ) - ( val < T( 0 ) ); }
+
+/* Stable symmetric Householder reflection that gives c and s such that
+ *   [ c  s ][a] = [d],
+ *   [ s -c ][b]   [0]
+ *
+ * Output:
+ *   c -- cosine(theta), where theta is the implicit angle of rotation
+ *        (counter-clockwise) in a plane-rotation
+ *   s -- sine(theta)
+ *   r -- two-norm of [a; b]
+ */
+static inline void symOrtho( double a, double b, double &c, double &s, double &r )
+{
+  if ( b == 0 )
+  {
+    c = mathSign( a );
+    s = 0;
+    r = std::abs( a );
+  }
+  else if ( a == 0 )
+  {
+    c = 0;
+    s = mathSign( b );
+    r = std::abs( b );
+  }
+  else if ( std::abs( b ) > std::abs( a ) )
+  {
+    const double tau = a / b;
+    s = mathSign( b ) / sqrt( 1 + tau * tau );
+    c = s * tau;
+    r = b / s;
+  }
+  else
+  {
+    const double tau = b / a;
+    c = mathSign( a ) / sqrt( 1 + tau * tau );
+    s = c * tau;
+    r = a / c;
+  }
+}
+
+/* Iterative LSQR algorithm for solving least squares problems.
+ *
+ * [1] Paige, C. C. and M. A. Saunders,
+ * LSQR: An Algorithm for Sparse Linear Equations And Sparse Least Squares
+ * ACM Trans. Math. Soft., Vol.8, 1982, pp. 43-71.
+ *
+ * Solves the following problem:
+ *   argmin_x ||Ax - b|| + damp||x||
+ *
+ * Output:
+ *   x -- approximate solution
+ */
+static void solveLSQR( const Mat &A, const Mat &b, OutputArray xOut, const double damp = 0.0,
```

Comment:
What is the best way to add a unit test for internal function? Can't find any example in the existing opencv code.